src/replication.c

```diff
@@ -2622,6 +3248,142 @@ int slaveTryPartialResynchronization(connection *conn, int read_reply) {
     return PSYNC_NOT_SUPPORTED;
 }
 
+/* Replication: Replica side.
+ * This connection handler fires after rdb-channel was initialized. We use it
+ * to adjust the replica main for loading incremental changes into the local buffer. */
+void setupMainConnForPsync(connection *conn) {
+    int psync_result = -1;
+    char llstr[LONG_STR_SIZE];
+    char* err = NULL;
+    if (server.repl_state == REPL_STATE_SEND_HANDSHAKE) {
+        /* We already have an initialized connection at master side, we only need to associate it with RDB connection */
+        ll2string(llstr,sizeof(llstr), server.rdb_client_id);
+        err = sendCommand(conn, "REPLCONF", "set-rdb-conn-id", llstr, NULL);
+        if (err) goto error;
+        server.repl_state = REPL_STATE_RECEIVE_CAPA_REPLY;
+        return;
+    }
+
+    if (server.repl_state == REPL_STATE_RECEIVE_CAPA_REPLY) {
+        err = receiveSynchronousResponse(conn);
+        if (err == NULL) goto error;
+        if (err[0] == '-') {
+            serverLog(LL_NOTICE,"Master does not understand REPLCONF identify: %s", err);
+            goto error;
+        }
+        sdsfree(err);
+        err = NULL;
+        server.repl_state = REPL_STATE_SEND_PSYNC;
+    }
+
+    if (server.repl_state == REPL_STATE_SEND_PSYNC) {
+        if (server.debug_sleep_after_fork) usleep(server.debug_sleep_after_fork);
+        if (slaveTryPartialResynchronization(conn,0) == PSYNC_WRITE_ERROR) {
+            serverLog(LL_WARNING, "Aborting RDB connection sync. Write error.");
+            cancelReplicationHandshake(1);
+        }
+        server.repl_state = REPL_STATE_RECEIVE_PSYNC_REPLY;
+        return;
+    }
+    psync_result = slaveTryPartialResynchronization(conn,1);
+    if (psync_result == PSYNC_WAIT_REPLY) return; /* Try again later... */
+
+    if (psync_result == PSYNC_CONTINUE) {
+        serverLog(LL_NOTICE, "MASTER <-> REPLICA sync: Master accepted a Partial Resynchronization%s",
+            server.repl_rdb_transfer_s != NULL ? ", RDB load in background.":".");
+        if (server.supervised_mode == SUPERVISED_SYSTEMD) {
+            serverCommunicateSystemd("STATUS=MASTER <-> REPLICA sync: Partial Resynchronization accepted. Ready to accept connections in read-write mode.\n");
+        }
+        completeTaskRDBChannelSyncMainConn(conn);
+        return;
+    }
+
+    error:
+    /* The rdb-conn-sync session must be aborted for any psync_result other than PSYNC_CONTINUE or PSYNC_WAIT_REPLY. */
+    serverLog(LL_WARNING, "Aborting RDB connection sync. Main connection psync result %d", psync_result);
+    cancelReplicationHandshake(1);
+}
+
+/*
```

Comment:
Thanks for the documentation! Couple of comments

1. first of all, can you please fix the terminology? we now use `primary` and `replica` exclusively (with the exception of existing externally facing strings)

2. this PR introduces a lot more states and commands between the primary and the replica. Some of them are not very easy to grasp, such as `CLIENT_PROTECTED_RDB_CHANNEL`. It will be really helpful if you could capture them somewhere, either in the PR itself or on the github issue.

Reply:
>For the few remaining instances where we still use "master-slave" terms, I believe replacing them with "primary-replica" might lead to confusion.

I don't think this should be the case but I will take a look